lib/srv/discovery/discovery.go

```diff
@@ -756,13 +768,79 @@ func (s *Server) handleEC2Instances(instances *server.EC2Instances) error {
 	// instances.Rotation is true whenever the instances received need
 	// to be rotated, we don't want to filter out existing OpenSSH nodes as
 	// they all need to have the command run on them
-	if !instances.Rotation {
+	//
+	// Integration/EICE Nodes don't have heartbeat.
+	// Those Nodes must not be filtered, so that we can extend their expiration and sync labels.
+	if !instances.Rotation && instances.Integration == "" {
 		s.filterExistingEC2Nodes(instances)
 	}
 	if len(instances.Instances) == 0 {
 		return trace.NotFound("all fetched nodes already enrolled")
 	}
 
+	switch {
+	case instances.Integration != "":
+		s.handleEC2UsingEICE(instances)
+	default:
+		if err := s.handleEC2RemoteInstallation(instances); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	if err := s.emitUsageEvents(instances.MakeEvents()); err != nil {
+		s.Log.WithError(err).Debug("Error emitting usage event.")
+	}
+
+	return nil
+}
+
+// handleEC2UsingEICE creates a Teleport Node (EICE subkind) from the instances list.
+func (s *Server) handleEC2UsingEICE(instances *server.EC2Instances) {
+	awsInfo := &types.AWSInfo{
+		AccountID:   instances.AccountID,
+		Region:      instances.Region,
+		Integration: instances.Integration,
+	}
+
+	existingEICENodes := s.existingEICENodes()
+
+	// Add EC2 Instances using EICE method
+	for _, ec2Instance := range instances.Instances {
+		eiceNode, err := services.NewAWSNodeFromEC2v1Instance(ec2Instance.OriginalInstance, awsInfo)
+		if err != nil {
+			s.Log.WithError(err).WithField("instance_id", ec2Instance.InstanceID).Warn("Error converting to Teleport EICE Node")
+			continue
+		}
+		eiceNodeExpiration := s.clock.Now().Add(s.jitter(serverExpirationDuration))
+		eiceNode.SetExpiry(eiceNodeExpiration)
+
+		// Does the node exist already?
+		for _, existingEICENode := range existingEICENodes {
+			match := types.MatchLabels(existingEICENode, map[string]string{
+				types.AWSAccountIDLabel:  instances.AccountID,
+				types.AWSInstanceIDLabel: ec2Instance.InstanceID,
+			})
+			if match {
+				// Re-use the same Name so that `UpsertNode` replaces the node.
+				eiceNode.SetName(existingEICENode.GetName())
+				break
+			}
+		}
+
+		if _, err := s.AccessPoint.UpsertNode(s.ctx, eiceNode); err != nil {
```

Comment:
So this will upsert the full node each time. With our regular heartbeat we have 2 modes: "keepalive" that just updates the node's TTL so it doesn't expire, and full "upsert" when node's properties change. Can we do it similarly for the discovery service? Upserting full node each time can be expensive esp. when there's a lot of them.

I wonder if we can adapt https://github.com/gravitational/teleport/blob/master/lib/srv/heartbeat.go to use here as it already implements the required logic for comparing resources and doing keepalive vs upsert.

Reply:
We also need to update the node's labels.
Otherwise, they get out of sync and RBAC stops applying.

We could change the heartbeat to include the labels as well :thinking:
However, it would probably make it much slower.
I'm not familiar with the performance differences between just updating the TTL vs updating TTL+Labels vs updating the entire object.
Given this is an Upsert, I would expect it to not be super slow (if it was a conditional update, it would be slower for sure).